From c6a141b68d527d9febeb5f45964a6ce8e3648002 Mon Sep 17 00:00:00 2001
From: Xaver Hugl <xaver.hugl@kde.org>
Date: Mon, 17 Feb 2025 16:25:09 +0100
Subject: [PATCH 1/4] increase required wayland-protocols version to 1.41

This version contains the color management protocol.

This commit also adjusts the build-wayland script to mention
that the DEBIAN_BASE_TAG also has to be bumped.

Signed-off-by: Xaver Hugl <xaver.hugl@kde.org>
Reviewed-by: Eric Engestrom <eric@igalia.com>
Reviewed-by: Sebastian Wick <sebastian.wick@redhat.com>
Acked-by: Daniel Stone <daniels@collabora.com>
---
 .gitlab-ci/container/build-wayland.sh |  3 ++-
 .gitlab-ci/image-tags.yml             | 16 ++++++++--------
 meson.build                           |  2 +-
 subprojects/wayland-protocols.wrap    |  6 +++---
 4 files changed, 14 insertions(+), 13 deletions(-)

diff --git a/.gitlab-ci/container/build-wayland.sh b/.gitlab-ci/container/build-wayland.sh
index 446367bc8f681..551edcfe7dbcb 100644
--- a/.gitlab-ci/container/build-wayland.sh
+++ b/.gitlab-ci/container/build-wayland.sh
@@ -8,6 +8,7 @@ uncollapsed_section_start wayland "Building Wayland"
 # When changing this file, you need to bump the following
 # .gitlab-ci/image-tags.yml tags:
 # ALPINE_X86_64_BUILD_TAG
+# DEBIAN_BASE_TAG
 # DEBIAN_BUILD_TAG
 # DEBIAN_TEST_ANDROID_TAG
 # DEBIAN_TEST_GL_TAG
@@ -16,7 +17,7 @@ uncollapsed_section_start wayland "Building Wayland"
 # KERNEL_ROOTFS_TAG
 
 export LIBWAYLAND_VERSION="1.21.0"
-export WAYLAND_PROTOCOLS_VERSION="1.38"
+export WAYLAND_PROTOCOLS_VERSION="1.41"
 
 git clone https://gitlab.freedesktop.org/wayland/wayland
 cd wayland
diff --git a/.gitlab-ci/image-tags.yml b/.gitlab-ci/image-tags.yml
index f24826e892824..743e4cf592fee 100644
--- a/.gitlab-ci/image-tags.yml
+++ b/.gitlab-ci/image-tags.yml
@@ -13,10 +13,10 @@
 
 variables:
    DEBIAN_X86_64_BUILD_BASE_IMAGE: "debian/x86_64_build-base"
-   DEBIAN_BASE_TAG: "20250220-kernel"
+   DEBIAN_BASE_TAG: "20250223-way-prot"
 
    DEBIAN_X86_64_BUILD_IMAGE_PATH: "debian/x86_64_build"
-   DEBIAN_BUILD_TAG: "20250210-android-drm"
+   DEBIAN_BUILD_TAG: "20250223-way-prot"
 
    DEBIAN_X86_64_TEST_BASE_IMAGE: "debian/x86_64_test-base"
    DEBIAN_ARM32_TEST_BASE_IMAGE: "debian/arm32_test-base"
@@ -30,17 +30,17 @@ variables:
    DEBIAN_ARM64_TEST_IMAGE_VK_PATH: "debian/arm64_test-vk"
    DEBIAN_X86_64_TEST_ANDROID_IMAGE_PATH: "debian/x86_64_test-android"
 
-   DEBIAN_TEST_ANDROID_TAG: "20250212-up-angle"
-   DEBIAN_TEST_GL_TAG: "20250221-piglit-04"
-   DEBIAN_TEST_VK_TAG: "20250221-piglit-04"
-   KERNEL_ROOTFS_TAG: "20250221-mtk-kernel"
+   DEBIAN_TEST_ANDROID_TAG: "20250223-way-prot"
+   DEBIAN_TEST_GL_TAG: "20250223-way-prot"
+   DEBIAN_TEST_VK_TAG: "20250223-way-prot"
+   KERNEL_ROOTFS_TAG: "20250223-way-prot"
 
    DEBIAN_PYUTILS_IMAGE: "debian/x86_64_pyutils"
    DEBIAN_PYUTILS_TAG: "20250129-lavacli"
 
-   ALPINE_X86_64_BUILD_TAG: "20250221-wlproto"
+   ALPINE_X86_64_BUILD_TAG: "20250223-way-prot"
    ALPINE_X86_64_LAVA_SSH_TAG: "20250124-spirv-tools"
-   FEDORA_X86_64_BUILD_TAG: "20250128-llversionm"
+   FEDORA_X86_64_BUILD_TAG: "20250223-way-prot"
 
    KERNEL_TAG: "v6.13-rc4-mesa-50ca"
    KERNEL_REPO: "gfx-ci/linux"
diff --git a/meson.build b/meson.build
index dfae7eada45c4..24e174f61e96b 100644
--- a/meson.build
+++ b/meson.build
@@ -2064,7 +2064,7 @@ if with_platform_wayland
   else
     wl_scanner_arg = 'code'
   endif
-  dep_wl_protocols = dependency('wayland-protocols', version : '>= 1.38', default_options: [ 'tests=false' ])
+  dep_wl_protocols = dependency('wayland-protocols', version : '>= 1.41', default_options: [ 'tests=false' ])
   dep_wayland_client = dependency('wayland-client', version : '>=1.18')
   dep_wayland_server = dependency('wayland-server', version : '>=1.18')
   if with_egl
diff --git a/subprojects/wayland-protocols.wrap b/subprojects/wayland-protocols.wrap
index 61f306bdffda2..c82fece01e380 100644
--- a/subprojects/wayland-protocols.wrap
+++ b/subprojects/wayland-protocols.wrap
@@ -1,7 +1,7 @@
 [wrap-file]
-directory = wayland-protocols-1.38
-source_url = https://gitlab.freedesktop.org/wayland/wayland-protocols/-/releases/1.38/downloads/wayland-protocols-1.38.tar.xz
-source_filename = wayland-protocols-1.38.tar.xz
+directory = wayland-protocols-1.41
+source_url = https://gitlab.freedesktop.org/wayland/wayland-protocols/-/releases/1.41/downloads/wayland-protocols-1.41.tar.xz
+source_filename = wayland-protocols-1.41.tar.xz
 source_hash = ff17292c05159d2b20ce6cacfe42d7e31a28198fa1429a769b03af7c38581dbe
 
 [provide]
-- 
GitLab


From dfaface1d57316aff29c9f94a5127d8a297ea4a1 Mon Sep 17 00:00:00 2001
From: Colin Marc <hi@colinmarc.com>
Date: Tue, 26 Mar 2024 11:12:28 +0100
Subject: [PATCH 2/4] vulkan/wsi: implement the Wayland color management
 protocol

This allows applications to use color spaces other than sRGB, if the compositor
supports them.

The color management surface is only created if a non-sRGB and non-passthrough
colorspace is set on the surface, so applications can still use the protocol
directly if they prefer.

Co-authored-by: Xaver Hugl <xaver.hugl@kde.org>
Reviewed-by: Sebastian Wick <sebastian.wick@redhat.com>
Acked-by: Daniel Stone <daniels@collabora.com>
---
 src/egl/wayland/wayland-drm/meson.build |   1 +
 src/vulkan/wsi/meson.build              |   1 +
 src/vulkan/wsi/wsi_common_wayland.c     | 421 ++++++++++++++++++++++--
 3 files changed, 400 insertions(+), 23 deletions(-)

diff --git a/src/egl/wayland/wayland-drm/meson.build b/src/egl/wayland/wayland-drm/meson.build
index 59ce34424999a..d3b6c15ddd210 100644
--- a/src/egl/wayland/wayland-drm/meson.build
+++ b/src/egl/wayland/wayland-drm/meson.build
@@ -48,6 +48,7 @@ wp_protos = {
   'presentation-time': 'stable/presentation-time/presentation-time.xml',
   'tearing-control-v1': 'staging/tearing-control/tearing-control-v1.xml',
   'linux-drm-syncobj-v1': 'staging/linux-drm-syncobj/linux-drm-syncobj-v1.xml',
+  'color-management-v1': 'staging/color-management/color-management-v1.xml',
 }
 wp_files = {}
 foreach name, xml : wp_protos
diff --git a/src/vulkan/wsi/meson.build b/src/vulkan/wsi/meson.build
index 6933eb780c40b..ba609ab3151db 100644
--- a/src/vulkan/wsi/meson.build
+++ b/src/vulkan/wsi/meson.build
@@ -22,6 +22,7 @@ if with_platform_wayland
   files_vulkan_wsi += wp_files['tearing-control-v1']
   links_vulkan_wsi += libloader_wayland_helper
   files_vulkan_wsi += wp_files['linux-drm-syncobj-v1']
+  files_vulkan_wsi += wp_files['color-management-v1']
 endif
 
 if with_platform_windows
diff --git a/src/vulkan/wsi/wsi_common_wayland.c b/src/vulkan/wsi/wsi_common_wayland.c
index 1f6176756b8f2..1036f4886f8de 100644
--- a/src/vulkan/wsi/wsi_common_wayland.c
+++ b/src/vulkan/wsi/wsi_common_wayland.c
@@ -47,6 +47,7 @@
 #include "presentation-time-client-protocol.h"
 #include "linux-drm-syncobj-v1-client-protocol.h"
 #include "tearing-control-v1-client-protocol.h"
+#include "color-management-v1-client-protocol.h"
 
 #include <util/cnd_monotonic.h>
 #include <util/compiler.h>
@@ -110,8 +111,13 @@ struct wsi_wl_display {
    struct wp_tearing_control_manager_v1 *tearing_control_manager;
    struct wp_linux_drm_syncobj_manager_v1 *wl_syncobj;
 
+   struct wp_color_manager_v1 *color_manager;
+
    struct dmabuf_feedback_format_table format_table;
 
+   struct u_vector color_primaries;
+   struct u_vector color_transfer_funcs;
+
    /* users want per-chain wsi_wl_swapchain->present_ids.wp_presentation */
    struct wp_presentation *wp_presentation_notwrapped;
    uint32_t wp_presentation_version;
@@ -125,6 +131,9 @@ struct wsi_wl_display {
    /* Formats populated by zwp_linux_dmabuf_v1 or wl_shm interfaces */
    struct u_vector formats;
 
+   /* Additional colorspaces returned by wp_color_management_v1. */
+   struct u_vector colorspaces;
+
    bool sw;
 
    dev_t main_device;
@@ -180,6 +189,11 @@ struct wsi_wl_surface {
    struct dmabuf_feedback dmabuf_feedback, pending_dmabuf_feedback;
 
    struct wp_linux_drm_syncobj_surface_v1 *wl_syncobj_surface;
+
+   struct vk_instance *instance;
+   struct wp_color_management_surface_v1 *color_surface;
+   int color_surface_refcount;
+   VkColorSpaceKHR colorspace;
 };
 
 struct wsi_wl_swapchain {
@@ -230,6 +244,8 @@ struct wsi_wl_swapchain {
       unsigned int refresh_nsec;
    } present_ids;
 
+   VkColorSpaceKHR colorspace;
+
    struct wsi_wl_image images[0];
 };
 VK_DEFINE_NONDISP_HANDLE_CASTS(wsi_wl_swapchain, base.base, VkSwapchainKHR,
@@ -885,6 +901,316 @@ static const struct wl_shm_listener shm_listener = {
    .format = shm_handle_format
 };
 
+static bool
+vector_contains(struct u_vector *vec, unsigned int val)
+{
+   unsigned int *ptr;
+
+   u_vector_foreach(ptr, vec)
+      if (*ptr == val)
+         return true;
+
+   return false;
+}
+
+struct Colorspace {
+   VkColorSpaceKHR colorspace;
+   enum wp_color_manager_v1_primaries primaries;
+   enum wp_color_manager_v1_transfer_function tf;
+};
+struct Colorspace colorspace_mapping[] = {
+   {
+      .colorspace = VK_COLOR_SPACE_SRGB_NONLINEAR_KHR,
+      .primaries = WP_COLOR_MANAGER_V1_PRIMARIES_SRGB,
+      .tf = WP_COLOR_MANAGER_V1_TRANSFER_FUNCTION_SRGB,
+   },
+   {
+      .colorspace = VK_COLOR_SPACE_DISPLAY_P3_NONLINEAR_EXT,
+      .primaries = WP_COLOR_MANAGER_V1_PRIMARIES_DISPLAY_P3,
+      .tf = WP_COLOR_MANAGER_V1_TRANSFER_FUNCTION_SRGB,
+   },
+   {
+      .colorspace = VK_COLOR_SPACE_EXTENDED_SRGB_LINEAR_EXT,
+      .primaries = WP_COLOR_MANAGER_V1_PRIMARIES_SRGB,
+      .tf = WP_COLOR_MANAGER_V1_TRANSFER_FUNCTION_EXT_LINEAR,
+   },
+   {
+      .colorspace = VK_COLOR_SPACE_DISPLAY_P3_LINEAR_EXT,
+      .primaries = WP_COLOR_MANAGER_V1_PRIMARIES_DISPLAY_P3,
+      .tf = WP_COLOR_MANAGER_V1_TRANSFER_FUNCTION_EXT_LINEAR,
+   },
+   {
+      .colorspace = VK_COLOR_SPACE_BT709_LINEAR_EXT,
+      .primaries = WP_COLOR_MANAGER_V1_PRIMARIES_SRGB,
+      .tf = WP_COLOR_MANAGER_V1_TRANSFER_FUNCTION_EXT_LINEAR,
+   },
+   {
+      .colorspace = VK_COLOR_SPACE_BT709_NONLINEAR_EXT,
+      .primaries = WP_COLOR_MANAGER_V1_PRIMARIES_SRGB,
+      .tf = WP_COLOR_MANAGER_V1_TRANSFER_FUNCTION_BT1886,
+   },
+   {
+      .colorspace = VK_COLOR_SPACE_BT2020_LINEAR_EXT,
+      .primaries = WP_COLOR_MANAGER_V1_PRIMARIES_BT2020,
+      .tf = WP_COLOR_MANAGER_V1_TRANSFER_FUNCTION_EXT_LINEAR,
+   },
+   {
+      .colorspace = VK_COLOR_SPACE_HDR10_ST2084_EXT,
+      .primaries = WP_COLOR_MANAGER_V1_PRIMARIES_BT2020,
+      .tf = WP_COLOR_MANAGER_V1_TRANSFER_FUNCTION_ST2084_PQ,
+   },
+   /* VK_COLOR_SPACE_DOLBYVISION_EXT is left out because it's deprecated */
+   {
+      .colorspace = VK_COLOR_SPACE_HDR10_HLG_EXT,
+      .primaries = WP_COLOR_MANAGER_V1_PRIMARIES_BT2020,
+      .tf = WP_COLOR_MANAGER_V1_TRANSFER_FUNCTION_HLG,
+   },
+   {
+      .colorspace = VK_COLOR_SPACE_ADOBERGB_LINEAR_EXT,
+      .primaries = WP_COLOR_MANAGER_V1_PRIMARIES_ADOBE_RGB,
+      .tf = WP_COLOR_MANAGER_V1_TRANSFER_FUNCTION_EXT_LINEAR,
+   },
+   /* VK_COLOR_SPACE_ADOBERGB_NONLINEAR_EXT is left out because there's no
+    * exactly matching transfer function in the Wayland protocol */
+   /* VK_COLOR_SPACE_PASS_THROUGH_EXT is handled elsewhere */
+   {
+      .colorspace = VK_COLOR_SPACE_EXTENDED_SRGB_NONLINEAR_EXT,
+      .primaries = WP_COLOR_MANAGER_V1_PRIMARIES_SRGB,
+      .tf = WP_COLOR_MANAGER_V1_TRANSFER_FUNCTION_EXT_SRGB,
+   },
+   /* VK_COLOR_SPACE_DISPLAY_NATIVE_AMD isn't supported */
+   /* VK_COLORSPACE_SRGB_NONLINEAR_KHR is just an alias */
+   /* VK_COLOR_SPACE_DCI_P3_LINEAR_EXT is just an alias */
+};
+
+static int
+wsi_wl_display_determine_colorspaces(struct wsi_wl_display *display)
+{
+   u_vector_finish(&display->colorspaces);
+   if (!u_vector_init(&display->colorspaces, 8, sizeof(VkColorSpaceKHR)))
+      return -1;
+
+   /* SRGB_NONLINEAR is always supported. */
+   VkColorSpaceKHR *new_cs = u_vector_add(&display->colorspaces);
+   if (!new_cs)
+      return -1;
+   *new_cs = VK_COLOR_SPACE_SRGB_NONLINEAR_KHR;
+
+   /* as is PASS_THROUGH */
+   new_cs = u_vector_add(&display->colorspaces);
+   if (!new_cs)
+      return -1;
+   *new_cs = VK_COLOR_SPACE_PASS_THROUGH_EXT;
+
+   if (!display->color_manager)
+      return 0;
+
+   struct u_vector *tfs = &display->color_transfer_funcs;
+   struct u_vector *primaries = &display->color_primaries;
+   for (int i = 0; i < ARRAY_SIZE(colorspace_mapping); i++) {
+      if (!vector_contains(primaries, colorspace_mapping[i].primaries))
+         continue;
+      if (!vector_contains(tfs, colorspace_mapping[i].tf))
+         continue;
+      VkColorSpaceKHR *new_cs = u_vector_add(&display->colorspaces);
+      if (!new_cs)
+         return -1;
+      *new_cs = colorspace_mapping[i].colorspace;
+   }
+   return 0;
+}
+
+static void
+color_management_handle_supported_intent(void *data,
+                                         struct wp_color_manager_v1 *color_manager,
+                                         unsigned int intent)
+{
+   /* We only use the perceptual rendering intent, which is always supported. */
+}
+
+static void
+color_management_handle_supported_features(void *data,
+                                           struct wp_color_manager_v1 *color_manager,
+                                           unsigned int feature)
+{
+   /* We don't use any non-default features yet. */
+}
+
+static void
+color_management_handle_supported_tf_named(void *data,
+                                           struct wp_color_manager_v1 *color_manager,
+                                           unsigned int tf)
+{
+   struct wsi_wl_display *display = data;
+   unsigned int *new_tf = u_vector_add(&display->color_transfer_funcs);
+   if (new_tf)
+      *new_tf = tf;
+}
+
+static void
+color_management_handle_supported_primaries_named(void *data,
+                                                  struct wp_color_manager_v1 *color_manager,
+                                                  unsigned int primaries)
+{
+   struct wsi_wl_display *display = data;
+   unsigned int *new_primaries = u_vector_add(&display->color_primaries);
+   if (new_primaries)
+      *new_primaries = primaries;
+}
+
+static void
+color_management_handle_done(void *data, struct wp_color_manager_v1 *color_manager)
+{
+   /* Intentionally left blank */
+}
+
+static const struct wp_color_manager_v1_listener color_manager_listener = {
+   .supported_intent = color_management_handle_supported_intent,
+   .supported_feature = color_management_handle_supported_features,
+   .supported_tf_named = color_management_handle_supported_tf_named,
+   .supported_primaries_named = color_management_handle_supported_primaries_named,
+   .done = color_management_handle_done,
+};
+
+enum image_description_status {
+   undefined,
+   ready,
+   failed,
+};
+
+static void
+color_management_handle_image_desc_failed(void *data,
+                                          struct wp_image_description_v1 *desc,
+                                          unsigned int cause,
+                                          const char *msg)
+{
+   enum image_description_status *status = data;
+   *status = failed;
+}
+
+static void
+color_management_handle_image_desc_ready(void *data,
+                                        struct wp_image_description_v1 *desc,
+                                        unsigned int id)
+{
+   enum image_description_status *status = data;
+   *status = ready;
+}
+
+static const struct wp_image_description_v1_listener image_description_listener = {
+   .failed = color_management_handle_image_desc_failed,
+   .ready = color_management_handle_image_desc_ready,
+};
+
+static bool
+needs_color_surface(VkColorSpaceKHR colorspace)
+{
+   return colorspace != VK_COLOR_SPACE_PASS_THROUGH_EXT;
+}
+
+static void
+wsi_wl_surface_add_color_refcount(struct wsi_wl_surface *wsi_surface)
+{
+   wsi_surface->color_surface_refcount++;
+   if (wsi_surface->color_surface_refcount == 1) {
+      wsi_surface->color_surface =
+         wp_color_manager_v1_get_surface(wsi_surface->display->color_manager, wsi_surface->surface);
+   }
+}
+
+static void
+wsi_wl_surface_remove_color_refcount(struct wsi_wl_surface *wsi_surface)
+{
+   wsi_surface->color_surface_refcount--;
+   if (wsi_surface->color_surface_refcount == 0) {
+      wp_color_management_surface_v1_destroy(wsi_surface->color_surface);
+      wsi_surface->color_surface = NULL;
+   }
+}
+
+static VkResult
+wsi_wl_swapchain_update_colorspace(struct wsi_wl_swapchain *chain)
+{
+   struct wsi_wl_surface *surface = chain->wsi_wl_surface;
+   struct wsi_wl_display *display = surface->display;
+
+   /* we need the color management extension for
+    * everything except sRGB and PASS_THROUGH */
+   if (!display->color_manager) {
+      if (chain->colorspace == VK_COLOR_SPACE_SRGB_NONLINEAR_KHR ||
+          chain->colorspace == VK_COLOR_SPACE_PASS_THROUGH_EXT) {
+         return VK_SUCCESS;
+      } else {
+         return VK_ERROR_SURFACE_LOST_KHR;
+      }
+   }
+
+   bool new_color_surface = !surface->color_surface;
+   bool needs_color_surface_new = needs_color_surface(chain->colorspace);
+   bool needs_color_surface_old = needs_color_surface(surface->colorspace);
+   if ((new_color_surface || !needs_color_surface_old) && needs_color_surface_new) {
+      wsi_wl_surface_add_color_refcount(surface);
+   } else if (needs_color_surface_old && !needs_color_surface_new) {
+      wsi_wl_surface_remove_color_refcount(surface);
+   }
+
+   if (!new_color_surface && surface->colorspace == chain->colorspace)
+      return VK_SUCCESS;
+
+   /* failure is fatal, so this potentially being wrong
+      in that case doesn't matter */
+   surface->colorspace = chain->colorspace;
+   if (!needs_color_surface_new)
+      return VK_SUCCESS;
+
+   struct wp_image_description_creator_params_v1 *creator =
+      wp_color_manager_v1_create_parametric_creator(display->color_manager);
+
+   if (!creator)
+      return VK_ERROR_SURFACE_LOST_KHR;
+
+   unsigned int primaries = 0;
+   unsigned int tf = 0;
+   for (int i = 0; i < ARRAY_SIZE(colorspace_mapping); i++) {
+      if (colorspace_mapping[i].colorspace == chain->colorspace) {
+         primaries = colorspace_mapping[i].primaries;
+         tf = colorspace_mapping[i].tf;
+      }
+   }
+
+   if (!primaries)
+      return VK_ERROR_SURFACE_LOST_KHR;
+
+   wp_image_description_creator_params_v1_set_primaries_named(creator, primaries);
+   wp_image_description_creator_params_v1_set_tf_named(creator, tf);
+
+   wl_proxy_set_queue((struct wl_proxy *) creator, display->queue);
+
+   struct wp_image_description_v1 *image_desc =
+      wp_image_description_creator_params_v1_create(creator);
+   if (!image_desc)
+      return VK_ERROR_SURFACE_LOST_KHR;
+
+   enum image_description_status status = undefined;
+   wp_image_description_v1_add_listener(image_desc, &image_description_listener, &status);
+
+   while (status == undefined) {
+      int ret = wl_display_dispatch_queue(display->wl_display, display->queue);
+      if (ret < 0)
+         return VK_ERROR_OUT_OF_DATE_KHR;
+   }
+   if (status == failed)
+      return VK_ERROR_SURFACE_LOST_KHR;
+
+   wp_color_management_surface_v1_set_image_description(chain->wsi_wl_surface->color_surface,
+                                                        image_desc,
+                                                        WP_COLOR_MANAGER_V1_RENDER_INTENT_PERCEPTUAL);
+   wp_image_description_v1_destroy(image_desc);
+
+   return VK_SUCCESS;
+}
+
+
 static void
 presentation_handle_clock_id(void* data, struct wp_presentation *wp_presentation, uint32_t clk_id)
 {
@@ -942,6 +1268,17 @@ registry_handle_global(void *data, struct wl_registry *registry,
       display->commit_timing_manager =
          wl_registry_bind(registry, name, &wp_commit_timing_manager_v1_interface, 1);
    }
+
+   if (strcmp(interface, wp_color_manager_v1_interface.name) == 0) {
+      display->color_manager =
+         wl_registry_bind(registry, name, &wp_color_manager_v1_interface, 1);
+
+      u_vector_init(&display->color_primaries, 8, sizeof(uint32_t));
+      u_vector_init(&display->color_transfer_funcs, 8, sizeof(uint32_t));
+
+      wp_color_manager_v1_add_listener(display->color_manager,
+                                       &color_manager_listener, display);
+   }
 }
 
 static void
@@ -961,6 +1298,10 @@ wsi_wl_display_finish(struct wsi_wl_display *display)
    u_vector_foreach(f, &display->formats)
       u_vector_finish(&f->modifiers);
    u_vector_finish(&display->formats);
+   u_vector_finish(&display->colorspaces);
+   u_vector_finish(&display->color_primaries);
+   u_vector_finish(&display->color_transfer_funcs);
+
    if (display->wl_shm)
       wl_shm_destroy(display->wl_shm);
    if (display->wl_syncobj)
@@ -975,6 +1316,8 @@ wsi_wl_display_finish(struct wsi_wl_display *display)
       wp_commit_timing_manager_v1_destroy(display->commit_timing_manager);
    if (display->tearing_control_manager)
       wp_tearing_control_manager_v1_destroy(display->tearing_control_manager);
+   if (display->color_manager)
+      wp_color_manager_v1_destroy(display->color_manager);
    if (display->wl_display_wrapper)
       wl_proxy_wrapper_destroy(display->wl_display_wrapper);
    if (display->queue)
@@ -1066,9 +1409,14 @@ wsi_wl_display_init(struct wsi_wayland *wsi_wl,
          }
    }
 
-   /* Round-trip again to get formats and modifiers */
+   /* Round-trip again to get formats, modifiers and color properties */
    wl_display_roundtrip_queue(display->wl_display, display->queue);
 
+   if (wsi_wl_display_determine_colorspaces(display) < 0) {
+      result = VK_ERROR_OUT_OF_HOST_MEMORY;
+      goto fail;
+   }
+
    if (wsi_wl->wsi->force_bgra8_unorm_first) {
       /* Find BGRA8_UNORM in the list and swap it to the first position if we
        * can find it.  Some apps get confused if SRGB is first in the list.
@@ -1359,6 +1707,8 @@ wsi_wl_surface_get_formats(VkIcdSurfaceBase *icd_surface,
                            VkSurfaceFormatKHR* pSurfaceFormats)
 {
    VkIcdSurfaceWayland *surface = (VkIcdSurfaceWayland *)icd_surface;
+   struct wsi_wl_surface *wsi_wl_surface =
+      wl_container_of((VkIcdSurfaceWayland *)icd_surface, wsi_wl_surface, base);
    struct wsi_wayland *wsi =
       (struct wsi_wayland *)wsi_device->wsi[VK_ICD_WSI_PLATFORM_WAYLAND];
 
@@ -1370,18 +1720,22 @@ wsi_wl_surface_get_formats(VkIcdSurfaceBase *icd_surface,
    VK_OUTARRAY_MAKE_TYPED(VkSurfaceFormatKHR, out,
                           pSurfaceFormats, pSurfaceFormatCount);
 
-   struct wsi_wl_format *disp_fmt;
-   u_vector_foreach(disp_fmt, &display.formats) {
-      /* Skip formats for which we can't support both alpha & opaque
-       * formats.
-       */
-      if (!(disp_fmt->flags & WSI_WL_FMT_ALPHA) ||
-          !(disp_fmt->flags & WSI_WL_FMT_OPAQUE))
-         continue;
+   VkColorSpaceKHR *cs;
+   u_vector_foreach(cs, &display.colorspaces) {
+      struct wsi_wl_format *disp_fmt;
+      u_vector_foreach(disp_fmt, &display.formats) {
+         /* Skip formats for which we can't support both alpha & opaque
+          * formats.
+          */
+         if (!(disp_fmt->flags & WSI_WL_FMT_ALPHA) ||
+            !(disp_fmt->flags & WSI_WL_FMT_OPAQUE)) {
+            continue;
+         }
 
-      vk_outarray_append_typed(VkSurfaceFormatKHR, &out, out_fmt) {
-         out_fmt->format = disp_fmt->vk_format;
-         out_fmt->colorSpace = VK_COLOR_SPACE_SRGB_NONLINEAR_KHR;
+         vk_outarray_append_typed(VkSurfaceFormatKHR, &out, out_fmt) {
+            out_fmt->format = disp_fmt->vk_format;
+            out_fmt->colorSpace = *cs;
+         }
       }
    }
 
@@ -1398,6 +1752,8 @@ wsi_wl_surface_get_formats2(VkIcdSurfaceBase *icd_surface,
                             VkSurfaceFormat2KHR* pSurfaceFormats)
 {
    VkIcdSurfaceWayland *surface = (VkIcdSurfaceWayland *)icd_surface;
+   struct wsi_wl_surface *wsi_wl_surface =
+      wl_container_of((VkIcdSurfaceWayland *)icd_surface, wsi_wl_surface, base);
    struct wsi_wayland *wsi =
       (struct wsi_wayland *)wsi_device->wsi[VK_ICD_WSI_PLATFORM_WAYLAND];
 
@@ -1409,18 +1765,22 @@ wsi_wl_surface_get_formats2(VkIcdSurfaceBase *icd_surface,
    VK_OUTARRAY_MAKE_TYPED(VkSurfaceFormat2KHR, out,
                           pSurfaceFormats, pSurfaceFormatCount);
 
-   struct wsi_wl_format *disp_fmt;
-   u_vector_foreach(disp_fmt, &display.formats) {
-      /* Skip formats for which we can't support both alpha & opaque
-       * formats.
-       */
-      if (!(disp_fmt->flags & WSI_WL_FMT_ALPHA) ||
-          !(disp_fmt->flags & WSI_WL_FMT_OPAQUE))
-         continue;
+   VkColorSpaceKHR *cs;
+   u_vector_foreach(cs, &display.colorspaces) {
+      struct wsi_wl_format *disp_fmt;
+      u_vector_foreach(disp_fmt, &display.formats) {
+         /* Skip formats for which we can't support both alpha & opaque
+          * formats.
+          */
+         if (!(disp_fmt->flags & WSI_WL_FMT_ALPHA) ||
+            !(disp_fmt->flags & WSI_WL_FMT_OPAQUE)) {
+            continue;
+         }
 
-      vk_outarray_append_typed(VkSurfaceFormat2KHR, &out, out_fmt) {
-         out_fmt->surfaceFormat.format = disp_fmt->vk_format;
-         out_fmt->surfaceFormat.colorSpace = VK_COLOR_SPACE_SRGB_NONLINEAR_KHR;
+         vk_outarray_append_typed(VkSurfaceFormat2KHR, &out, out_fmt) {
+            out_fmt->surfaceFormat.format = disp_fmt->vk_format;
+            out_fmt->surfaceFormat.colorSpace = *cs;
+         }
       }
    }
 
@@ -1516,6 +1876,9 @@ wsi_wl_surface_destroy(VkIcdSurfaceBase *icd_surface, VkInstance _instance,
       dmabuf_feedback_fini(&wsi_wl_surface->pending_dmabuf_feedback);
    }
 
+   if (wsi_wl_surface->color_surface)
+      wp_color_management_surface_v1_destroy(wsi_wl_surface->color_surface);
+
    if (wsi_wl_surface->surface)
       wl_proxy_wrapper_destroy(wsi_wl_surface->surface);
 
@@ -1838,6 +2201,9 @@ wsi_CreateWaylandSurfaceKHR(VkInstance _instance,
    surface->display = pCreateInfo->display;
    surface->surface = pCreateInfo->surface;
 
+   wsi_wl_surface->instance = instance;
+   wsi_wl_surface->colorspace = VK_COLOR_SPACE_SRGB_NONLINEAR_KHR;
+
    *pSurface = VkIcdSurfaceBase_to_handle(&surface->base);
 
    return VK_SUCCESS;
@@ -2460,6 +2826,10 @@ wsi_wl_swapchain_queue_present(struct wsi_swapchain *wsi_chain,
              image->base.row_pitches[0] * chain->extent.height);
    }
 
+   VkResult ret = wsi_wl_swapchain_update_colorspace(chain);
+   if (ret != VK_SUCCESS)
+      return ret;
+
    /* For EXT_swapchain_maintenance1. We might have transitioned from FIFO to MAILBOX.
     * In this case we need to let the FIFO request complete, before presenting MAILBOX. */
    while (!chain->legacy_fifo_ready) {
@@ -2816,6 +3186,9 @@ wsi_wl_swapchain_chain_free(struct wsi_wl_swapchain *chain,
       wl_callback_destroy(chain->frame);
    if (chain->tearing_control)
       wp_tearing_control_v1_destroy(chain->tearing_control);
+   if (needs_color_surface(chain->colorspace) && wsi_wl_surface->color_surface) {
+      wsi_wl_surface_remove_color_refcount(wsi_wl_surface);
+   }
 
    /* Only unregister if we are the non-retired swapchain, or
     * we are a retired swapchain and memory allocation failed,
@@ -2980,6 +3353,8 @@ wsi_wl_surface_create_swapchain(VkIcdSurfaceBase *icd_surface,
                                                           WP_TEARING_CONTROL_V1_PRESENTATION_HINT_ASYNC);
    }
 
+   chain->colorspace = pCreateInfo->imageColorSpace;
+
    enum wsi_wl_buffer_type buffer_type;
    struct wsi_base_image_params *image_params = NULL;
    struct wsi_cpu_image_params cpu_image_params;
-- 
GitLab


From b7540ae22d591e6d3166f07e1afd72696a5cdb2e Mon Sep 17 00:00:00 2001
From: Xaver Hugl <xaver.hugl@kde.org>
Date: Tue, 5 Nov 2024 19:17:28 +0100
Subject: [PATCH 3/4] vulkan/wsi: implement support for VK_EXT_hdr_metadata on
 Wayland

Signed-off-by: Xaver Hugl <xaver.hugl@kde.org>
Reviewed-by: Alyssa Rosenzweig <alyssa@rosenzweig.io>
Reviewed-by: Sebastian Wick <sebastian.wick@redhat.com>
Acked-by: Daniel Stone <daniels@collabora.com>
---
 src/amd/vulkan/radv_physical_device.c       |   1 +
 src/asahi/vulkan/hk_physical_device.c       |   1 +
 src/freedreno/vulkan/tu_device.cc           |   1 +
 src/gallium/frontends/lavapipe/lvp_device.c |   1 +
 src/intel/vulkan/anv_physical_device.c      |   1 +
 src/nouveau/vulkan/nvk_physical_device.c    |   1 +
 src/panfrost/vulkan/panvk_physical_device.c |   1 +
 src/vulkan/wsi/wsi_common.c                 |  12 ++
 src/vulkan/wsi/wsi_common_private.h         |   2 +
 src/vulkan/wsi/wsi_common_wayland.c         | 151 ++++++++++++++++----
 10 files changed, 146 insertions(+), 26 deletions(-)

diff --git a/src/amd/vulkan/radv_physical_device.c b/src/amd/vulkan/radv_physical_device.c
index 98ae3028c51c2..79526ca3eadc6 100644
--- a/src/amd/vulkan/radv_physical_device.c
+++ b/src/amd/vulkan/radv_physical_device.c
@@ -684,6 +684,7 @@ radv_physical_device_get_supported_extensions(const struct radv_physical_device
       .EXT_global_priority = true,
       .EXT_global_priority_query = true,
       .EXT_graphics_pipeline_library = !pdev->use_llvm && !(instance->debug_flags & RADV_DEBUG_NO_GPL),
+      .EXT_hdr_metadata = true,
       .EXT_host_query_reset = true,
       .EXT_image_2d_view_of_3d = true,
       .EXT_image_compression_control = true,
diff --git a/src/asahi/vulkan/hk_physical_device.c b/src/asahi/vulkan/hk_physical_device.c
index b45a69f714102..c5334a0d67659 100644
--- a/src/asahi/vulkan/hk_physical_device.c
+++ b/src/asahi/vulkan/hk_physical_device.c
@@ -153,6 +153,7 @@ hk_get_device_extensions(const struct hk_instance *instance,
       .EXT_global_priority = true,
       .EXT_global_priority_query = true,
       .EXT_graphics_pipeline_library = true,
+      .EXT_hdr_metadata = true,
       .EXT_host_query_reset = true,
       .EXT_host_image_copy = true,
       .EXT_image_2d_view_of_3d = true,
diff --git a/src/freedreno/vulkan/tu_device.cc b/src/freedreno/vulkan/tu_device.cc
index af6d854569afa..a73ca0d9a76aa 100644
--- a/src/freedreno/vulkan/tu_device.cc
+++ b/src/freedreno/vulkan/tu_device.cc
@@ -274,6 +274,7 @@ get_device_extensions(const struct tu_physical_device *device,
       .EXT_global_priority = true,
       .EXT_global_priority_query = true,
       .EXT_graphics_pipeline_library = true,
+      .EXT_hdr_metadata = true,
       .EXT_host_image_copy = true,
       .EXT_host_query_reset = true,
       .EXT_image_2d_view_of_3d = true,
diff --git a/src/gallium/frontends/lavapipe/lvp_device.c b/src/gallium/frontends/lavapipe/lvp_device.c
index 55877dd68de8e..17461689e4a97 100644
--- a/src/gallium/frontends/lavapipe/lvp_device.c
+++ b/src/gallium/frontends/lavapipe/lvp_device.c
@@ -212,6 +212,7 @@ static const struct vk_device_extension_table lvp_device_extensions_supported =
    .EXT_extended_dynamic_state3           = true,
    .EXT_external_memory_host              = true,
    .EXT_graphics_pipeline_library         = true,
+   .EXT_hdr_metadata = true,
    .EXT_host_image_copy                   = true,
    .EXT_host_query_reset                  = true,
    .EXT_image_2d_view_of_3d               = true,
diff --git a/src/intel/vulkan/anv_physical_device.c b/src/intel/vulkan/anv_physical_device.c
index 37a20edcef1fd..d4243b886b76f 100644
--- a/src/intel/vulkan/anv_physical_device.c
+++ b/src/intel/vulkan/anv_physical_device.c
@@ -282,6 +282,7 @@ get_device_extensions(const struct anv_physical_device *device,
       .EXT_global_priority_query             = device->max_context_priority >=
                                                VK_QUEUE_GLOBAL_PRIORITY_MEDIUM_KHR,
       .EXT_graphics_pipeline_library         = !debug_get_bool_option("ANV_NO_GPL", false),
+      .EXT_hdr_metadata = true,
       .EXT_host_image_copy                   = !device->emu_astc_ldr,
       .EXT_host_query_reset                  = true,
       .EXT_image_2d_view_of_3d               = true,
diff --git a/src/nouveau/vulkan/nvk_physical_device.c b/src/nouveau/vulkan/nvk_physical_device.c
index 4dd9d8b7b3376..d3d80491639ef 100644
--- a/src/nouveau/vulkan/nvk_physical_device.c
+++ b/src/nouveau/vulkan/nvk_physical_device.c
@@ -228,6 +228,7 @@ nvk_get_device_extensions(const struct nvk_instance *instance,
       .EXT_global_priority = true,
       .EXT_global_priority_query = true,
       .EXT_graphics_pipeline_library = true,
+      .EXT_hdr_metadata = true,
       .EXT_host_query_reset = true,
       .EXT_host_image_copy = info->cls_eng3d >= TURING_A,
       .EXT_image_2d_view_of_3d = true,
diff --git a/src/panfrost/vulkan/panvk_physical_device.c b/src/panfrost/vulkan/panvk_physical_device.c
index 67ac92fdf7b6f..3a03332a5e649 100644
--- a/src/panfrost/vulkan/panvk_physical_device.c
+++ b/src/panfrost/vulkan/panvk_physical_device.c
@@ -237,6 +237,7 @@ get_device_extensions(const struct panvk_physical_device *device,
       .EXT_global_priority = true,
       .EXT_global_priority_query = true,
       .EXT_graphics_pipeline_library = true,
+      .EXT_hdr_metadata = true,
       .EXT_host_query_reset = true,
       .EXT_image_drm_format_modifier = true,
       .EXT_image_robustness = true,
diff --git a/src/vulkan/wsi/wsi_common.c b/src/vulkan/wsi/wsi_common.c
index f1dea2344edfd..97b19b454f098 100644
--- a/src/vulkan/wsi/wsi_common.c
+++ b/src/vulkan/wsi/wsi_common.c
@@ -2265,3 +2265,15 @@ wsi_device_supports_explicit_sync(struct wsi_device *device)
       (device->timeline_semaphore_export_handle_types &
        VK_EXTERNAL_SEMAPHORE_HANDLE_TYPE_OPAQUE_FD_BIT);
 }
+
+VKAPI_ATTR void VKAPI_CALL
+wsi_SetHdrMetadataEXT(VkDevice device, uint32_t swapchainCount,
+                      const VkSwapchainKHR* pSwapchains,
+                      const VkHdrMetadataEXT* pMetadata)
+{
+   for (uint32_t i = 0; i < swapchainCount; i++) {
+      VK_FROM_HANDLE(wsi_swapchain, swapchain, pSwapchains[i]);
+      if (swapchain->set_hdr_metadata)
+         swapchain->set_hdr_metadata(swapchain, pMetadata);
+   }
+}
diff --git a/src/vulkan/wsi/wsi_common_private.h b/src/vulkan/wsi/wsi_common_private.h
index 23981db703c0b..51bfa77d4b2f7 100644
--- a/src/vulkan/wsi/wsi_common_private.h
+++ b/src/vulkan/wsi/wsi_common_private.h
@@ -225,6 +225,8 @@ struct wsi_swapchain {
                               const uint32_t *indices);
    void (*set_present_mode)(struct wsi_swapchain *swap_chain,
                             VkPresentModeKHR mode);
+   void (*set_hdr_metadata)(struct wsi_swapchain *swap_chain,
+                            const VkHdrMetadataEXT* pMetadata);
 };
 
 bool
diff --git a/src/vulkan/wsi/wsi_common_wayland.c b/src/vulkan/wsi/wsi_common_wayland.c
index 1036f4886f8de..ee5eabb56ce5c 100644
--- a/src/vulkan/wsi/wsi_common_wayland.c
+++ b/src/vulkan/wsi/wsi_common_wayland.c
@@ -140,6 +140,11 @@ struct wsi_wl_display {
    bool same_gpu;
 
    clockid_t presentation_clock_id;
+
+   struct {
+      bool mastering_display_primaries;
+      bool extended_target_volume;
+   } color_features;
 };
 
 struct wsi_wayland {
@@ -191,9 +196,14 @@ struct wsi_wl_surface {
    struct wp_linux_drm_syncobj_surface_v1 *wl_syncobj_surface;
 
    struct vk_instance *instance;
-   struct wp_color_management_surface_v1 *color_surface;
-   int color_surface_refcount;
-   VkColorSpaceKHR colorspace;
+
+   struct {
+      struct wp_color_management_surface_v1 *color_surface;
+      int color_surface_refcount;
+      VkColorSpaceKHR colorspace;
+      VkHdrMetadataEXT hdr_metadata;
+      bool has_hdr_metadata;
+   } color;
 };
 
 struct wsi_wl_swapchain {
@@ -244,7 +254,11 @@ struct wsi_wl_swapchain {
       unsigned int refresh_nsec;
    } present_ids;
 
-   VkColorSpaceKHR colorspace;
+   struct {
+      VkColorSpaceKHR colorspace;
+      VkHdrMetadataEXT hdr_metadata;
+      bool has_hdr_metadata;
+   } color;
 
    struct wsi_wl_image images[0];
 };
@@ -917,58 +931,69 @@ struct Colorspace {
    VkColorSpaceKHR colorspace;
    enum wp_color_manager_v1_primaries primaries;
    enum wp_color_manager_v1_transfer_function tf;
+   bool should_use_hdr_metadata;
 };
 struct Colorspace colorspace_mapping[] = {
    {
       .colorspace = VK_COLOR_SPACE_SRGB_NONLINEAR_KHR,
       .primaries = WP_COLOR_MANAGER_V1_PRIMARIES_SRGB,
       .tf = WP_COLOR_MANAGER_V1_TRANSFER_FUNCTION_SRGB,
+      .should_use_hdr_metadata = false,
    },
    {
       .colorspace = VK_COLOR_SPACE_DISPLAY_P3_NONLINEAR_EXT,
       .primaries = WP_COLOR_MANAGER_V1_PRIMARIES_DISPLAY_P3,
       .tf = WP_COLOR_MANAGER_V1_TRANSFER_FUNCTION_SRGB,
+      .should_use_hdr_metadata = false,
    },
    {
       .colorspace = VK_COLOR_SPACE_EXTENDED_SRGB_LINEAR_EXT,
       .primaries = WP_COLOR_MANAGER_V1_PRIMARIES_SRGB,
       .tf = WP_COLOR_MANAGER_V1_TRANSFER_FUNCTION_EXT_LINEAR,
+      .should_use_hdr_metadata = true,
    },
    {
       .colorspace = VK_COLOR_SPACE_DISPLAY_P3_LINEAR_EXT,
       .primaries = WP_COLOR_MANAGER_V1_PRIMARIES_DISPLAY_P3,
       .tf = WP_COLOR_MANAGER_V1_TRANSFER_FUNCTION_EXT_LINEAR,
+      .should_use_hdr_metadata = false,
    },
    {
       .colorspace = VK_COLOR_SPACE_BT709_LINEAR_EXT,
       .primaries = WP_COLOR_MANAGER_V1_PRIMARIES_SRGB,
       .tf = WP_COLOR_MANAGER_V1_TRANSFER_FUNCTION_EXT_LINEAR,
+      .should_use_hdr_metadata = false,
    },
    {
       .colorspace = VK_COLOR_SPACE_BT709_NONLINEAR_EXT,
       .primaries = WP_COLOR_MANAGER_V1_PRIMARIES_SRGB,
       .tf = WP_COLOR_MANAGER_V1_TRANSFER_FUNCTION_BT1886,
+      .should_use_hdr_metadata = false,
    },
    {
       .colorspace = VK_COLOR_SPACE_BT2020_LINEAR_EXT,
       .primaries = WP_COLOR_MANAGER_V1_PRIMARIES_BT2020,
       .tf = WP_COLOR_MANAGER_V1_TRANSFER_FUNCTION_EXT_LINEAR,
+      .should_use_hdr_metadata = false,
    },
    {
       .colorspace = VK_COLOR_SPACE_HDR10_ST2084_EXT,
       .primaries = WP_COLOR_MANAGER_V1_PRIMARIES_BT2020,
       .tf = WP_COLOR_MANAGER_V1_TRANSFER_FUNCTION_ST2084_PQ,
+      .should_use_hdr_metadata = true,
    },
    /* VK_COLOR_SPACE_DOLBYVISION_EXT is left out because it's deprecated */
    {
       .colorspace = VK_COLOR_SPACE_HDR10_HLG_EXT,
       .primaries = WP_COLOR_MANAGER_V1_PRIMARIES_BT2020,
       .tf = WP_COLOR_MANAGER_V1_TRANSFER_FUNCTION_HLG,
+      .should_use_hdr_metadata = true,
    },
    {
       .colorspace = VK_COLOR_SPACE_ADOBERGB_LINEAR_EXT,
       .primaries = WP_COLOR_MANAGER_V1_PRIMARIES_ADOBE_RGB,
       .tf = WP_COLOR_MANAGER_V1_TRANSFER_FUNCTION_EXT_LINEAR,
+      .should_use_hdr_metadata = false,
    },
    /* VK_COLOR_SPACE_ADOBERGB_NONLINEAR_EXT is left out because there's no
     * exactly matching transfer function in the Wayland protocol */
@@ -977,6 +1002,7 @@ struct Colorspace colorspace_mapping[] = {
       .colorspace = VK_COLOR_SPACE_EXTENDED_SRGB_NONLINEAR_EXT,
       .primaries = WP_COLOR_MANAGER_V1_PRIMARIES_SRGB,
       .tf = WP_COLOR_MANAGER_V1_TRANSFER_FUNCTION_EXT_SRGB,
+      .should_use_hdr_metadata = true,
    },
    /* VK_COLOR_SPACE_DISPLAY_NATIVE_AMD isn't supported */
    /* VK_COLORSPACE_SRGB_NONLINEAR_KHR is just an alias */
@@ -1033,7 +1059,17 @@ color_management_handle_supported_features(void *data,
                                            struct wp_color_manager_v1 *color_manager,
                                            unsigned int feature)
 {
-   /* We don't use any non-default features yet. */
+   struct wsi_wl_display *display = data;
+   switch (feature) {
+   case WP_COLOR_MANAGER_V1_FEATURE_SET_MASTERING_DISPLAY_PRIMARIES:
+      display->color_features.mastering_display_primaries = true;
+      break;
+   case WP_COLOR_MANAGER_V1_FEATURE_EXTENDED_TARGET_VOLUME:
+      display->color_features.extended_target_volume = true;
+      break;
+   default:
+      break;
+   }
 }
 
 static void
@@ -1111,9 +1147,9 @@ needs_color_surface(VkColorSpaceKHR colorspace)
 static void
 wsi_wl_surface_add_color_refcount(struct wsi_wl_surface *wsi_surface)
 {
-   wsi_surface->color_surface_refcount++;
-   if (wsi_surface->color_surface_refcount == 1) {
-      wsi_surface->color_surface =
+   wsi_surface->color.color_surface_refcount++;
+   if (wsi_surface->color.color_surface_refcount == 1) {
+      wsi_surface->color.color_surface =
          wp_color_manager_v1_get_surface(wsi_surface->display->color_manager, wsi_surface->surface);
    }
 }
@@ -1121,13 +1157,30 @@ wsi_wl_surface_add_color_refcount(struct wsi_wl_surface *wsi_surface)
 static void
 wsi_wl_surface_remove_color_refcount(struct wsi_wl_surface *wsi_surface)
 {
-   wsi_surface->color_surface_refcount--;
-   if (wsi_surface->color_surface_refcount == 0) {
-      wp_color_management_surface_v1_destroy(wsi_surface->color_surface);
-      wsi_surface->color_surface = NULL;
+   wsi_surface->color.color_surface_refcount--;
+   if (wsi_surface->color.color_surface_refcount == 0) {
+      wp_color_management_surface_v1_destroy(wsi_surface->color.color_surface);
+      wsi_surface->color.color_surface = NULL;
    }
 }
 
+static bool
+compare_hdr_metadata(struct VkHdrMetadataEXT *l, struct VkHdrMetadataEXT *r)
+{
+   return l->displayPrimaryRed.x == r->displayPrimaryRed.x
+       && l->displayPrimaryRed.y == r->displayPrimaryRed.y
+       && l->displayPrimaryGreen.x == r->displayPrimaryGreen.x
+       && l->displayPrimaryGreen.y == r->displayPrimaryGreen.y
+       && l->displayPrimaryBlue.x == r->displayPrimaryBlue.x
+       && l->displayPrimaryBlue.y == r->displayPrimaryBlue.y
+       && l->whitePoint.x == r->whitePoint.x
+       && l->whitePoint.y == r->whitePoint.y
+       && l->maxLuminance == r->maxLuminance
+       && l->minLuminance == r->minLuminance
+       && l->maxContentLightLevel == r->maxContentLightLevel
+       && l->maxFrameAverageLightLevel == r->maxFrameAverageLightLevel;
+}
+
 static VkResult
 wsi_wl_swapchain_update_colorspace(struct wsi_wl_swapchain *chain)
 {
@@ -1137,29 +1190,43 @@ wsi_wl_swapchain_update_colorspace(struct wsi_wl_swapchain *chain)
    /* we need the color management extension for
     * everything except sRGB and PASS_THROUGH */
    if (!display->color_manager) {
-      if (chain->colorspace == VK_COLOR_SPACE_SRGB_NONLINEAR_KHR ||
-          chain->colorspace == VK_COLOR_SPACE_PASS_THROUGH_EXT) {
+      if (chain->color.colorspace == VK_COLOR_SPACE_SRGB_NONLINEAR_KHR ||
+          chain->color.colorspace == VK_COLOR_SPACE_PASS_THROUGH_EXT) {
          return VK_SUCCESS;
       } else {
          return VK_ERROR_SURFACE_LOST_KHR;
       }
    }
 
-   bool new_color_surface = !surface->color_surface;
-   bool needs_color_surface_new = needs_color_surface(chain->colorspace);
-   bool needs_color_surface_old = needs_color_surface(surface->colorspace);
+   bool new_color_surface = !surface->color.color_surface;
+   bool needs_color_surface_new = needs_color_surface(chain->color.colorspace);
+   bool needs_color_surface_old = needs_color_surface(surface->color.colorspace);
    if ((new_color_surface || !needs_color_surface_old) && needs_color_surface_new) {
       wsi_wl_surface_add_color_refcount(surface);
    } else if (needs_color_surface_old && !needs_color_surface_new) {
       wsi_wl_surface_remove_color_refcount(surface);
    }
 
-   if (!new_color_surface && surface->colorspace == chain->colorspace)
+   bool should_use_hdr_metadata = chain->color.has_hdr_metadata;
+   for (int i = 0; i < ARRAY_SIZE(colorspace_mapping); i++) {
+      if (colorspace_mapping[i].colorspace == chain->color.colorspace) {
+         should_use_hdr_metadata &= colorspace_mapping[i].should_use_hdr_metadata;
+         break;
+      }
+   }
+
+   if (!new_color_surface &&
+       surface->color.colorspace == chain->color.colorspace &&
+       surface->color.has_hdr_metadata == should_use_hdr_metadata &&
+       compare_hdr_metadata(&surface->color.hdr_metadata, &chain->color.hdr_metadata)) {
       return VK_SUCCESS;
+   }
 
    /* failure is fatal, so this potentially being wrong
       in that case doesn't matter */
-   surface->colorspace = chain->colorspace;
+   surface->color.colorspace = chain->color.colorspace;
+   surface->color.hdr_metadata = chain->color.hdr_metadata;
+   surface->color.has_hdr_metadata = should_use_hdr_metadata;
    if (!needs_color_surface_new)
       return VK_SUCCESS;
 
@@ -1172,7 +1239,7 @@ wsi_wl_swapchain_update_colorspace(struct wsi_wl_swapchain *chain)
    unsigned int primaries = 0;
    unsigned int tf = 0;
    for (int i = 0; i < ARRAY_SIZE(colorspace_mapping); i++) {
-      if (colorspace_mapping[i].colorspace == chain->colorspace) {
+      if (colorspace_mapping[i].colorspace == chain->color.colorspace) {
          primaries = colorspace_mapping[i].primaries;
          tf = colorspace_mapping[i].tf;
       }
@@ -1183,6 +1250,29 @@ wsi_wl_swapchain_update_colorspace(struct wsi_wl_swapchain *chain)
 
    wp_image_description_creator_params_v1_set_primaries_named(creator, primaries);
    wp_image_description_creator_params_v1_set_tf_named(creator, tf);
+   if (should_use_hdr_metadata && display->color_features.extended_target_volume) {
+      uint32_t max_cll = round(chain->color.hdr_metadata.maxContentLightLevel);
+      uint32_t max_fall = round(chain->color.hdr_metadata.maxFrameAverageLightLevel);
+      wp_image_description_creator_params_v1_set_max_cll(creator, max_cll);
+      wp_image_description_creator_params_v1_set_max_fall(creator, max_fall);
+      if (display->color_features.mastering_display_primaries) {
+         uint32_t red_x = round(chain->color.hdr_metadata.displayPrimaryRed.x * 1000000);
+         uint32_t red_y = round(chain->color.hdr_metadata.displayPrimaryRed.y * 1000000);
+         uint32_t green_x = round(chain->color.hdr_metadata.displayPrimaryGreen.x * 1000000);
+         uint32_t green_y = round(chain->color.hdr_metadata.displayPrimaryGreen.y * 1000000);
+         uint32_t blue_x = round(chain->color.hdr_metadata.displayPrimaryBlue.x * 1000000);
+         uint32_t blue_y = round(chain->color.hdr_metadata.displayPrimaryBlue.y * 1000000);
+         uint32_t white_x = round(chain->color.hdr_metadata.whitePoint.x * 1000000);
+         uint32_t white_y = round(chain->color.hdr_metadata.whitePoint.y * 1000000);
+         wp_image_description_creator_params_v1_set_mastering_display_primaries(creator, red_x, red_y,
+                                                                                green_x, green_y,
+                                                                                blue_x, blue_y,
+                                                                                white_x, white_y);
+         uint32_t min_lum = round(chain->color.hdr_metadata.minLuminance * 10000);
+         uint32_t max_lum = round(chain->color.hdr_metadata.maxLuminance);
+         wp_image_description_creator_params_v1_set_mastering_luminance(creator, min_lum, max_lum);
+      }
+   }
 
    wl_proxy_set_queue((struct wl_proxy *) creator, display->queue);
 
@@ -1202,7 +1292,7 @@ wsi_wl_swapchain_update_colorspace(struct wsi_wl_swapchain *chain)
    if (status == failed)
       return VK_ERROR_SURFACE_LOST_KHR;
 
-   wp_color_management_surface_v1_set_image_description(chain->wsi_wl_surface->color_surface,
+   wp_color_management_surface_v1_set_image_description(chain->wsi_wl_surface->color.color_surface,
                                                         image_desc,
                                                         WP_COLOR_MANAGER_V1_RENDER_INTENT_PERCEPTUAL);
    wp_image_description_v1_destroy(image_desc);
@@ -1210,6 +1300,14 @@ wsi_wl_swapchain_update_colorspace(struct wsi_wl_swapchain *chain)
    return VK_SUCCESS;
 }
 
+static void
+wsi_wl_swapchain_set_hdr_metadata(struct wsi_swapchain *wsi_chain,
+                                  const VkHdrMetadataEXT* pMetadata)
+{
+   struct wsi_wl_swapchain *chain = (struct wsi_wl_swapchain *)wsi_chain;
+   chain->color.hdr_metadata = *pMetadata;
+   chain->color.has_hdr_metadata = true;
+}
 
 static void
 presentation_handle_clock_id(void* data, struct wp_presentation *wp_presentation, uint32_t clk_id)
@@ -1876,8 +1974,8 @@ wsi_wl_surface_destroy(VkIcdSurfaceBase *icd_surface, VkInstance _instance,
       dmabuf_feedback_fini(&wsi_wl_surface->pending_dmabuf_feedback);
    }
 
-   if (wsi_wl_surface->color_surface)
-      wp_color_management_surface_v1_destroy(wsi_wl_surface->color_surface);
+   if (wsi_wl_surface->color.color_surface)
+      wp_color_management_surface_v1_destroy(wsi_wl_surface->color.color_surface);
 
    if (wsi_wl_surface->surface)
       wl_proxy_wrapper_destroy(wsi_wl_surface->surface);
@@ -2202,7 +2300,7 @@ wsi_CreateWaylandSurfaceKHR(VkInstance _instance,
    surface->surface = pCreateInfo->surface;
 
    wsi_wl_surface->instance = instance;
-   wsi_wl_surface->colorspace = VK_COLOR_SPACE_SRGB_NONLINEAR_KHR;
+   wsi_wl_surface->color.colorspace = VK_COLOR_SPACE_SRGB_NONLINEAR_KHR;
 
    *pSurface = VkIcdSurfaceBase_to_handle(&surface->base);
 
@@ -3186,7 +3284,7 @@ wsi_wl_swapchain_chain_free(struct wsi_wl_swapchain *chain,
       wl_callback_destroy(chain->frame);
    if (chain->tearing_control)
       wp_tearing_control_v1_destroy(chain->tearing_control);
-   if (needs_color_surface(chain->colorspace) && wsi_wl_surface->color_surface) {
+   if (needs_color_surface(chain->color.colorspace) && wsi_wl_surface->color.color_surface) {
       wsi_wl_surface_remove_color_refcount(wsi_wl_surface);
    }
 
@@ -3353,7 +3451,7 @@ wsi_wl_surface_create_swapchain(VkIcdSurfaceBase *icd_surface,
                                                           WP_TEARING_CONTROL_V1_PRESENTATION_HINT_ASYNC);
    }
 
-   chain->colorspace = pCreateInfo->imageColorSpace;
+   chain->color.colorspace = pCreateInfo->imageColorSpace;
 
    enum wsi_wl_buffer_type buffer_type;
    struct wsi_base_image_params *image_params = NULL;
@@ -3427,6 +3525,7 @@ wsi_wl_surface_create_swapchain(VkIcdSurfaceBase *icd_surface,
    chain->base.wait_for_present = wsi_wl_swapchain_wait_for_present;
    chain->base.present_mode = present_mode;
    chain->base.image_count = num_images;
+   chain->base.set_hdr_metadata = wsi_wl_swapchain_set_hdr_metadata;
    chain->extent = pCreateInfo->imageExtent;
    chain->vk_format = pCreateInfo->imageFormat;
    chain->buffer_type = buffer_type;
-- 
GitLab


From 8d1d448e6624c3b427b47b7330a657e80778e5af Mon Sep 17 00:00:00 2001
From: Xaver Hugl <xaver.hugl@kde.org>
Date: Thu, 7 Nov 2024 13:58:49 +0100
Subject: [PATCH 4/4] vulkan/wsi: handle the compositor not supporting extended
 target volume better

Instead of unconditionally ignoring the HDR metadata, just attempt to create the image
description, and if it fails, fall back to creating it without HDR metadata.

Signed-off-by: Xaver Hugl <xaver.hugl@kde.org>
Reviewed-by: Sebastian Wick <sebastian.wick@redhat.com>
Acked-by: Daniel Stone <daniels@collabora.com>
---
 src/vulkan/wsi/wsi_common_wayland.c | 15 ++++++++++++---
 1 file changed, 12 insertions(+), 3 deletions(-)

diff --git a/src/vulkan/wsi/wsi_common_wayland.c b/src/vulkan/wsi/wsi_common_wayland.c
index ee5eabb56ce5c..098ba9c80928e 100644
--- a/src/vulkan/wsi/wsi_common_wayland.c
+++ b/src/vulkan/wsi/wsi_common_wayland.c
@@ -1250,7 +1250,7 @@ wsi_wl_swapchain_update_colorspace(struct wsi_wl_swapchain *chain)
 
    wp_image_description_creator_params_v1_set_primaries_named(creator, primaries);
    wp_image_description_creator_params_v1_set_tf_named(creator, tf);
-   if (should_use_hdr_metadata && display->color_features.extended_target_volume) {
+   if (should_use_hdr_metadata) {
       uint32_t max_cll = round(chain->color.hdr_metadata.maxContentLightLevel);
       uint32_t max_fall = round(chain->color.hdr_metadata.maxFrameAverageLightLevel);
       wp_image_description_creator_params_v1_set_max_cll(creator, max_cll);
@@ -1289,8 +1289,17 @@ wsi_wl_swapchain_update_colorspace(struct wsi_wl_swapchain *chain)
       if (ret < 0)
          return VK_ERROR_OUT_OF_DATE_KHR;
    }
-   if (status == failed)
-      return VK_ERROR_SURFACE_LOST_KHR;
+   if (status == failed) {
+      wp_image_description_v1_destroy(image_desc);
+      if (!display->color_features.extended_target_volume && should_use_hdr_metadata) {
+         /* VK_EXT_hdr_metadata doesn't specify if or how the metadata is used,
+          * so it's fine to try again without it. */
+         chain->color.has_hdr_metadata = false;
+         return wsi_wl_swapchain_update_colorspace(chain);
+      } else {
+         return VK_ERROR_SURFACE_LOST_KHR;
+      }
+   }
 
    wp_color_management_surface_v1_set_image_description(chain->wsi_wl_surface->color.color_surface,
                                                         image_desc,
-- 
GitLab

